### DISCUSSION TITLE
help request: plugin conflict handling and aprerequisites

### DISCUSSION THREAD

**Discussion by mygaochunming** [apache/apisix#7179] [Q&A] [answered]
I have read  the priority of the plugin by documents

[Plugin Loading Process](https://apisix.apache.org/docs/apisix/architecture-design/apisix#plugin-loading-process)

![image](https://user-images.githubusercontent.com/21029140/171319257-37d5f5a8-0d69-4dee-8fba-cceccc2d8676.png)

[plugin-config](https://apisix.apache.org/docs/apisix/architecture-design/plugin-config/)

![image](https://user-images.githubusercontent.com/21029140/171317820-db807d06-4190-4c91-826b-f90c3b57d602.png)

[plugin](https://apisix.apache.org/docs/apisix/terminology/plugin)

![image](https://user-images.githubusercontent.com/21029140/171318148-e5f19cdb-746f-4bcc-ae79-c1aea4413ce5.png)

the chapter [Global rules](https://apisix.apache.org/zh/docs/apisix/terminology/global-rule) does not tell me the conflict handling rules between "global rules" and rules in cosumer/route/service.

Exclude global rules, I think apisix handle plugin‘s flow like this(image 001)

![image](https://user-images.githubusercontent.com/21029140/171345456-bec376e3-118d-4c80-85ea-d43e31107291.png)

but when I read the plugins chapter ,  I'm confused .

**For Key-Auth plugin**

```
curl http://127.0.0.1:9080/apisix/admin/consumers -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "username": "jack",
    "plugins": {
        "key-auth": {
            "key": "auth-one"
        }
    }
}'

curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "methods": ["GET"],
    "uri": "/index.html",
    "id": 1,
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "127.0.0.1:1980": 1
        }
    }
}'



curl http://127.0.0.2:9080/index.html -i

=============I GOT BEGIN==================

HTTP/1.1 502 Bad Gateway
Date: Wed, 01 Jun 2022 07:13:05 GMT
Content-Type: text/html; charset=utf-8
Content-Length: 154
Connection: keep-alive
Server: APISIX/2.14.0
X-APISIX-Upstream-Status: 502

<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>openresty</center>
</body>
</html>
=============I GOT END==================

curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "methods": ["GET"],
    "uri": "/index.html",
    "id": 1,
    "plugins": {
        "key-auth": {}
    },
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "127.0.0.1:1980": 1
        }
    }
}'

curl http://127.0.0.2:9080/index.html -i

=============I GOT BEGIN==================

HTTP/1.1 401 Unauthorized
Date: Wed, 01 Jun 2022 07:16:45 GMT
Content-Type: text/html; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
Server: APISIX/2.14.0

{"message":"Missing API key found in request"}
=============I GOT END==================

curl http://127.0.0.2:9080/index.html -H 'apikey: auth-one' -i

=============I GOT BEGIN==================

HTTP/1.1 502 Bad Gateway
Date: Wed, 01 Jun 2022 07:17:49 GMT
Content-Type: text/html; charset=utf-8
Content-Length: 154
Connection: keep-alive
Server: APISIX/2.14.0
X-APISIX-Upstream-Status: 502

<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>openresty</center>
</body>
</html>
=============I GOT END==================

```

ignore the "502 Bad Gateway" error messages, It looks like that whether the Key-Auth plugin works or not is restricted by the route rule

```
"plugins": {
        "key-auth": {}
    }
```  

If it configed then the key-Auth works otherwise it not work.  

Maybe the operation of a plugin is not only affected by the priority and the above conflict handling rules, some plugins require aprerequisites to run.

Because I’m new about lua and openResty, it's hard for me to read the source code, I need help for explaining about :

![image](https://user-images.githubusercontent.com/21029140/171315029-9475b09f-8d2c-4c97-bb6e-4c4d9ca63666.png)

thanks in advance.

**Comment by moonming** [comment]
@membphis please take a look

**Reply by tokers** [reply]
Actually, the "plugin" you configured on the consumer is just some credentials, these "plugins" don't affect the running of APISIX. Whether a plugin will be executed for a request is decided by the current route matched with this request.

**Reply by mygaochunming** [reply]
Thanks very much.

**Reply by mygaochunming** [reply]
@tokers 
![image](https://user-images.githubusercontent.com/21029140/173482405-6b324755-b0c4-45f4-8dd8-6f822cca6971.png)
![image](https://user-images.githubusercontent.com/21029140/173483230-6bed508b-86f7-4d4b-ae9f-58df426495dd.png)
Does this mean the consumer's plugins will be run whether or not it defined in the route?

**Comment by membphis** [comment, marked as answer]
I drew a picture for you to know the detail, and you can get the flow from the source code: 

https://github.com/apache/apisix/blob/master/apisix/init.lua#L366-L453

<img width="1018" alt="image" src="https://user-images.githubusercontent.com/6814606/172553544-632c6031-3295-40cb-8a69-a5cb42f6cc11.png">

**Reply by mygaochunming** [reply]
Thanks for your answer, Much clearer now.
